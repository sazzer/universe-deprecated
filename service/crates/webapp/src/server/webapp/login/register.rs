use crate::server::request_id::RequestId;
use rocket::{post, request::LenientForm, FromForm};
use serde::Serialize;
use std::collections::HashMap;
use std::convert::{TryFrom, TryInto};
use universe_templates::Template;
use universe_users::{Password, UserData, Username};

/// The shape of the form data that the POST /login/register endpoing can accept
#[derive(FromForm, Debug, Serialize, Clone)]
pub struct RegisterForm {
    pub username: Option<String>,
    pub email: Option<String>,
    pub name: Option<String>,
    pub password: Option<String>,
    pub password2: Option<String>,
}

impl TryFrom<RegisterForm> for UserData {
    type Error = HashMap<&'static str, Option<&'static str>>;
    fn try_from(value: RegisterForm) -> Result<Self, Self::Error> {
        let passwords_match = value.password == value.password2;

        let username: Result<Username, &'static str> = value
            .username
            .ok_or("missing")
            .and_then(|u| u.parse().map_err(|_| "missing"));

        let email: Result<String, &'static str> = value
            .email
            .map(|s| s.trim().to_owned())
            .filter(|s| !s.is_empty())
            .ok_or("missing");
        let name: Result<String, &'static str> = value
            .name
            .map(|s| s.trim().to_owned())
            .filter(|s| !s.is_empty())
            .ok_or("missing");
        let password: Result<Password, &'static str> = value
            .password
            .ok_or("missing")
            .and_then(|p| Password::from_plaintext(p).map_err(|_| "hash_error"));
        let password2: Result<(), &'static str> = value.password2.ok_or("missing").and_then(|_| {
            if passwords_match {
                Ok(())
            } else {
                Err("different")
            }
        });

        match (username, email, name, password, password2) {
            (Ok(username), Ok(email), Ok(display_name), Ok(password), Ok(_)) => Ok(UserData {
                username,
                email,
                display_name,
                password,
            }),
            (username, email, name, password, password2) => {
                let mut errors = HashMap::new();
                errors.insert("username", username.err());
                errors.insert("email", email.err());
                errors.insert("name", name.err());
                errors.insert("password", password.err());
                errors.insert("password2", password2.err());
                Err(errors)
            }
        }
    }
}

#[post("/login/register", data = "<form>")]
#[tracing::instrument]
pub fn process_register(form: LenientForm<RegisterForm>, _request_id: RequestId) -> Template {
    let register_form = form.into_inner();
    let user: Result<UserData, HashMap<&str, Option<&str>>> = register_form.clone().try_into();

    let template = match user {
        Ok(user) => Template::new("login/register.tera"),
        Err(errors) => Template::new("login/register.tera").with_data("errors", &errors),
    };

    template
        .with_data("username", &register_form.username)
        .with_data("email", &register_form.email)
        .with_data("name", &register_form.name)
}

#[cfg(test)]
mod tests {
    use super::*;
    use spectral::prelude::*;

    #[test]
    fn test_convert_form_success() {
        let form = RegisterForm {
            username: Some("testuser".to_owned()),
            email: Some("test@example.com".to_owned()),
            name: Some("Test User".to_owned()),
            password: Some("pa55word".to_owned()),
            password2: Some("pa55word".to_owned()),
        };
        let user: UserData = form.try_into().unwrap();

        assert_that(&user.username).is_equal_to("testuser".parse::<Username>().unwrap());
        assert_that(&user.email).is_equal_to("test@example.com".to_owned());
        assert_that(&user.display_name).is_equal_to("Test User".to_owned());
        assert_that(&user.password.verify("pa55word")).is_equal_to(true);
    }

    #[test]
    fn test_convert_form_success_padded() {
        let form = RegisterForm {
            username: Some("  testuser  ".to_owned()),
            email: Some("  test@example.com  ".to_owned()),
            name: Some("  Test User  ".to_owned()),
            password: Some("  pa55word  ".to_owned()),
            password2: Some("  pa55word  ".to_owned()),
        };
        let user: UserData = form.try_into().unwrap();

        assert_that(&user.username).is_equal_to("testuser".parse::<Username>().unwrap());
        assert_that(&user.email).is_equal_to("test@example.com".to_owned());
        assert_that(&user.display_name).is_equal_to("Test User".to_owned());
        assert_that(&user.password.verify("  pa55word  ")).is_equal_to(true);
    }

    #[test]
    fn test_convert_form_all_null() {
        let form = RegisterForm {
            username: None,
            email: None,
            name: None,
            password: None,
            password2: None,
        };
        let user: Result<UserData, HashMap<&str, Option<&str>>> = form.try_into();
        let err = user.unwrap_err();

        assert_that(&err).has_length(5);
        assert_that(&err).contains_entry("username", Some("missing"));
        assert_that(&err).contains_entry("email", Some("missing"));
        assert_that(&err).contains_entry("name", Some("missing"));
        assert_that(&err).contains_entry("password", Some("missing"));
        assert_that(&err).contains_entry("password2", Some("missing"));
    }

    #[test]
    fn test_convert_form_all_spaces() {
        let form = RegisterForm {
            username: Some("    ".to_owned()),
            email: Some("    ".to_owned()),
            name: Some("    ".to_owned()),
            password: Some("    ".to_owned()),
            password2: Some("    ".to_owned()),
        };
        let user: Result<UserData, HashMap<&str, Option<&str>>> = form.try_into();
        let err = user.unwrap_err();

        assert_that(&err).has_length(5);
        assert_that(&err).contains_entry("username", Some("missing"));
        assert_that(&err).contains_entry("email", Some("missing"));
        assert_that(&err).contains_entry("name", Some("missing"));
        assert_that(&err).contains_entry("password", None);
        assert_that(&err).contains_entry("password2", None);
    }

    #[test]
    fn test_convert_form_password_mismatch() {
        let form = RegisterForm {
            username: Some("testuser".to_owned()),
            email: Some("test@example.com".to_owned()),
            name: Some("Test User".to_owned()),
            password: Some("pa55word".to_owned()),
            password2: Some("Pa55word".to_owned()),
        };
        let user: Result<UserData, HashMap<&str, Option<&str>>> = form.try_into();
        let err = user.unwrap_err();

        assert_that(&err).has_length(5);
        assert_that(&err).contains_entry("username", None);
        assert_that(&err).contains_entry("email", None);
        assert_that(&err).contains_entry("name", None);
        assert_that(&err).contains_entry("password", None);
        assert_that(&err).contains_entry("password2", Some("different"));
    }
}
