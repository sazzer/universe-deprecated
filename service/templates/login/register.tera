{% extends "layout/landing.tera" %}

{% block sidebar %}
<h3>{{t(key="login_title_register")}}</h3>
<form action="/login/register" method="POST">
  <div class="form-group">
    <label for="login_username">{{t(key="login_username_label")}}</label>
    <input type="text"
           class="form-control {% if errors.username is defined and errors.username is string %}is-invalid{% endif %}"
           id="login_username"
           name="username"
           value="{{ username | default(value="") | escape }}"
           readonly />
    {% if errors.username is defined and errors.username is string %}
    <div class="invalid-feedback">
      {{ t(key="login_username_problems_" ~ errors.username) }}
    </div>
    {% endif %}
  </div>
  <div class="form-group">
    <label for="login_email">{{t(key="login_email_label")}}</label>
    <input type="email"
           class="form-control {% if errors.email is defined and errors.email is string %}is-invalid{% endif %}"
           id="login_email"
           name="email"
           value="{{ email | default(value="") | escape }}"
           aria-describedby="login_email_help"
           autofocus
           required />
    {% if errors.email is defined and errors.email is string %}
    <div class="invalid-feedback">
      {{ t(key="login_email_problems_" ~ errors.email) }}
    </div>
    {% endif %}
    <small id="login_email_help" class="form-text text-muted">{{t(key="login_email_help")}}</small>
  </div>
  <div class="form-group">
    <label for="login_name">{{t(key="login_name_label")}}</label>
    <input type="text"
           class="form-control {% if errors.name is defined and errors.name is string %}is-invalid{% endif %}"
           id="login_name"
           name="name"
           value="{{ name | default(value="") | escape }}"
           required />
    {% if errors.name is defined and errors.name is string %}
    <div class="invalid-feedback">
      {{ t(key="login_name_problems_" ~ errors.name) }}
    </div>
    {% endif %}
  </div>
  <div class="form-group">
    <label for="login_password">{{t(key="login_password_label")}}</label>
    <input type="password"
           class="form-control {% if errors.password is defined and errors.password is string %}is-invalid{% endif %}"
           id="login_password"
           name="password"
           required />
    {% if errors.password is defined and errors.password is string %}
    <div class="invalid-feedback">
      {{ t(key="login_password_problems_" ~ errors.password) }}
    </div>
    {% endif %}
  </div>
  <div class="form-group">
    <label for="login_password2">{{t(key="login_password2_label")}}</label>
    <input type="password"
           class="form-control {% if errors.password2 is defined and errors.password2 is string %}is-invalid{% endif %}"
           id="login_password2"
           name="password2"
           required />
    {% if errors.password2 is defined and errors.password2 is string %}
    <div class="invalid-feedback">
      {{ t(key="login_password2_problems_" ~ errors.password2) }}
    </div>
    {% endif %}
  </div>
  <button type="submit" class="btn btn-primary">{{t(key="login_submit_register")}}</button>
</form>
{% endblock sidebar %}
