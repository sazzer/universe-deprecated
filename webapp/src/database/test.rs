use lazy_static::lazy_static;
use log::info;
use postgres::{Client, NoTls};
use testcontainers::{clients::Cli, images::postgres::Postgres, Container, Docker};

lazy_static! {
    static ref DOCKER: Cli = { Cli::default() };
}

pub struct TestDatabase<'d> {
    #[allow(dead_code)]
    node: Container<'d, Cli, Postgres>,
    url: String,
}

impl<'d> TestDatabase<'d> {
    pub fn new() -> Self {
        let node = DOCKER.run(Postgres::default());

        let host_port = node.get_host_port(5432).unwrap();
        let url = format!("postgres://postgres:postgres@localhost:{}", host_port);
        info!("Running postgres on {}", url);

        TestDatabase { node, url }
    }

    pub fn client(&self) -> Client {
        Client::connect(&self.url, NoTls).unwrap()
    }
}
