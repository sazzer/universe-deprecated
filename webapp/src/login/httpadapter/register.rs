use super::form::LoginForm;
use crate::users::{UserService, Username};
use log::info;
use rocket::request::LenientForm;
use std::collections::HashMap;
use universe_templates::Template;

/// Helper to see if an `Option<String>` actually has a string in it
///
/// # Arguments
/// * `input` The input to check
///
/// # Returns
/// False if the input is `None`, or if it is `Some(s)` where `s` is either empty or pure whitespace.
/// True if the input is `Some(s)` where `s` contains at least one non-whitespace character
fn is_empty(input: &Option<String>) -> bool {
    match input {
        None => false,
        Some(str) => str.trim().is_empty(),
    }
}

/// Handler to attempt to register a new user and react accordingly
///
/// # Arguments
/// * `body` The form submission that we are processing
/// * `username` The username that we are registering as
/// * `user_service` The user service to work with
///
/// TODO: Implement
pub fn perform_register(
    body: LenientForm<LoginForm>,
    username: Username,
    _user_service: &dyn UserService,
) -> Template {
    let mut errors: HashMap<&str, &str> = HashMap::new();

    if is_empty(&body.email) {
        errors.insert("email", "missing");
    }

    if is_empty(&body.name) {
        errors.insert("name", "missing");
    }

    if is_empty(&body.password) {
        errors.insert("password", "missing");
    }

    if is_empty(&body.password2) {
        errors.insert("password2", "missing");
    } else if body.password != body.password2 {
        errors.insert("password2", "different");
    }

    info!("Errors processing registration form: {:?}", errors);
    Template::new("login/register.tera")
        .with_data("username", username.as_ref())
        .with_data("email", body.email.as_ref().unwrap_or(&"".to_owned()))
        .with_data("name", body.name.as_ref().unwrap_or(&"".to_owned()))
        .with_data("errors", &errors)
}
