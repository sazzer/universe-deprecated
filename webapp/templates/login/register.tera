{% extends "layout/landing.tera" %}

{% block content %}
<div class="card">
  <div class="card-header">{{ t(key="login-title-register") }}</div>
  <div class="card-body">

    <form action="/login" method="POST">
      <input type="hidden" name="action" value="register" />
      <div class="form-group">
        <label for="username">{{ t(key="login-username-label") }}</label>
        <input type="username"
               class="form-control"
               id="username"
               name="username"
               value="{{username | trim}}"
               readonly />
      </div>
      <div class="form-group">
        <label for="email">{{ t(key="login-email-label") }}</label>
        <input type="email"
               class="form-control {% if errors.email %}is-invalid{%elif errors%}is-valid{%endif%}"
               id="email"
               name="email"
               value="{{email | default(value="") | trim}}"
               aria-describedby="emailHelp"
               autofocus
               required />
        <small id="emailHelp" class="form-text text-muted">{{ t(key="login-email-help")}}</small>
       {% if errors.email %}
       <div class="invalid-feedback">
         {{ t(key="login-email-errors-" ~ errors.email) }}
       </div>
       {% endif %}
      </div>
      <div class="form-group">
        <label for="name">{{ t(key="login-name-label") }}</label>
        <input type="text"
               class="form-control {% if errors.name %}is-invalid{%elif errors%}is-valid{%endif%}"
               id="name"
               name="name"
               value="{{name | default(value="") | trim}}"
               required />
        {% if errors.name %}
        <div class="invalid-feedback">
          {{ t(key="login-name-errors-" ~ errors.name) }}
        </div>
        {% endif %}
      </div>
      <div class="form-group">
        <label for="password">{{ t(key="login-password-label") }}</label>
        <input type="password"
               class="form-control {% if errors.password %}is-invalid{%elif errors%}is-valid{%endif%}"
               id="password"
               name="password"
               required />
        {% if errors.password %}
        <div class="invalid-feedback">
          {{ t(key="login-password-errors-" ~ errors.password) }}
        </div>
        {% endif %}
      </div>
      <div class="form-group">
        <label for="password2">{{ t(key="login-password2-label") }}</label>
        <input type="password"
               class="form-control {% if errors.password2 %}is-invalid{%elif errors%}is-valid{%endif%}"
               id="password2"
               name="password2"
               required />
        {% if errors.password2 %}
        <div class="invalid-feedback">
          {{ t(key="login-password2-errors-" ~ errors.password2) }}
        </div>
        {% endif %}
      </div>
      <button type="submit" class="btn btn-primary">{{ t(key="login-submit-register") }}</button>
      <a href="/login" class="btn btn-secondary">{{ t(key="login-submit-cancel") }}</a>
    </form>

  </div>
</div>
{% endblock %}
