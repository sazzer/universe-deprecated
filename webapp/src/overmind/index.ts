import { namespaced } from "overmind/config";
import { createHook } from "overmind-react";
import { IConfig } from "overmind";
import * as login from "../login";
import * as authentication from "../authentication";

export const config = namespaced({
  login,
  authentication
});

export const useOvermind = createHook<typeof config>();

declare module "overmind" {
  // tslint:disable:interface-name
  interface Config extends IConfig<typeof config> {}
}
