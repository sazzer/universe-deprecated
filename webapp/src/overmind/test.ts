import { createOvermindMock, OvermindMock, Config } from "overmind";
import { config } from "../overmind";
import { cloneDeep, merge } from "lodash";

/**
 * NestedPartial is stolen from Overmind to represent a part of the Overmind structure
 */
export declare type NestedPartial<T> = T extends Function
  ? T
  : Partial<
      {
        [P in keyof T]: NestedPartial<T[P]>;
      }
    >;

/**
 * Create a test instance of the Overmind to work with
 * @param mockEffects The mock effects to wire in
 * @param initialState The initial state to wire in
 */
export function createTestOvermind(
  mockEffects?: NestedPartial<Config["effects"]>,
  initialState?: NestedPartial<Config["state"]>
): OvermindMock<Config> {
  const mergedState = merge(cloneDeep(config), { state: initialState });
  return createOvermindMock(mergedState, mockEffects);
}
