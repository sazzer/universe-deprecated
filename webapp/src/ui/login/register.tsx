import React from "react";
import { useTranslation } from "react-i18next";
import { useForm, ErrorMessage, FieldValues } from "react-hook-form";
import * as yup from "yup";
import { useOvermind } from "../../overmind";

/**
 * Render the view for the Register Form
 */
export const RegisterForm: React.FC = () => {
  const { t } = useTranslation();
  const { state, actions } = useOvermind();

  const onSubmitHandler = async (data: FieldValues) => {
    await actions.login.register({
      username: data.username,
      email: data.email,
      displayName: data.displayName,
      password: data.password
    });
  };

  const { register, errors, handleSubmit } = useForm({
    validationSchema: yup.object().shape({
      username: yup
        .string()
        .required(t("login.username.errors.required"))
        .trim(),
      email: yup
        .string()
        .email(t("login.email.errors.email"))
        .required(t("login.email.errors.required"))
        .trim(),
      displayName: yup
        .string()
        .required(t("login.displayName.errors.required"))
        .trim(),
      password: yup.string().required(t("login.password.errors.required")),
      password2: yup
        .string()
        .required(t("login.password2.errors.required"))
        .when(["password"], (password: string, schema: any) => {
          return schema.oneOf(
            [password],
            t("login.password2.errors.different")
          );
        })
    }),
    validateCriteriaMode: "all",
    defaultValues: {
      username: state.login.username || "",
      email: "",
      displayName: "",
      password: "",
      password2: ""
    }
  });

  let errorMessage;

  return (
    <>
      <h3>{t("login.register.title")}</h3>
      <form onSubmit={handleSubmit(onSubmitHandler)} data-test="registerForm">
        <div className="form-group" data-test="username">
          <label htmlFor="username">{t("login.username.label")}</label>
          <input
            type="text"
            className={
              errors.username ? "form-control is-invalid" : "form-control"
            }
            id="username"
            name="username"
            readOnly
            ref={register}
          />
          <ErrorMessage
            errors={errors}
            name="username"
            as={<div className="invalid-feedback" />}
          />
        </div>
        <div className="form-group" data-test="email">
          <label htmlFor="email">{t("login.email.label")}</label>
          <input
            type="text"
            className={
              errors.email ? "form-control is-invalid" : "form-control"
            }
            id="email"
            name="email"
            autoFocus
            ref={register}
          />
          <ErrorMessage
            errors={errors}
            name="email"
            as={<div className="invalid-feedback" />}
          />
        </div>
        <div className="form-group" data-test="displayName">
          <label htmlFor="displayName">{t("login.displayName.label")}</label>
          <input
            type="text"
            className={
              errors.displayName ? "form-control is-invalid" : "form-control"
            }
            id="displayName"
            name="displayName"
            ref={register}
          />
          <ErrorMessage
            errors={errors}
            name="displayName"
            as={<div className="invalid-feedback" />}
          />
        </div>
        <div className="form-group" data-test="password">
          <label htmlFor="password">{t("login.password.label")}</label>
          <input
            type="password"
            className={
              errors.password ? "form-control is-invalid" : "form-control"
            }
            id="password"
            name="password"
            ref={register}
          />
          <ErrorMessage
            errors={errors}
            name="password"
            as={<div className="invalid-feedback" />}
          />
        </div>
        <div className="form-group" data-test="password2">
          <label htmlFor="password2">{t("login.password2.label")}</label>
          <input
            type="password"
            className={
              errors.password2 ? "form-control is-invalid" : "form-control"
            }
            id="password2"
            name="password2"
            ref={register}
          />
          <ErrorMessage
            errors={errors}
            name="password2"
            as={<div className="invalid-feedback" />}
          />
        </div>
        <div className="form-group">
          <button
            type="submit"
            className="btn btn-primary"
            disabled={state.login.loading}
          >
            {state.login.loading && (
              <span
                className="spinner-border spinner-border-sm"
                role="status"
                aria-hidden="true"
              ></span>
            )}
            {t("login.register.submit")}
          </button>
          <button
            type="button"
            className="btn btn-link"
            disabled={state.login.loading}
            onClick={actions.login.cancelLogin}
          >
            {t("login.register.cancel")}
          </button>
        </div>
        {errorMessage}
      </form>
    </>
  );
};
