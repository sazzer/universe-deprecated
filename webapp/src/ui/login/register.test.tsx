import React from "react";
import { render, wait, fireEvent } from "@testing-library/react";
import { Provider } from "overmind-react";
import { RegisterForm } from "./register";
import { createTestOvermind } from "../../overmind/test";
import { OvermindMock, Config } from "overmind";

describe("RegisterForm", () => {
  test("Initial render", () => {
    const overmind = createTestOvermind(
      {},
      {
        login: {
          mode: {
            current: "registering"
          },
          username: "testuser"
        }
      }
    );
    const { container } = render(
      <Provider value={overmind}>
        <RegisterForm />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
  test("Cancelling Registration", async () => {
    const overmind = createTestOvermind(
      {},
      {
        login: {
          mode: {
            current: "registering"
          },
          username: "testuser"
        }
      }
    );
    const { getByText } = render(
      <Provider value={overmind}>
        <RegisterForm />
      </Provider>
    );
    await wait(() =>
      fireEvent.click(getByText("Cancel", { selector: "button" }))
    );
    expect(overmind.mutations).toMatchSnapshot();
  });
  describe("Submitting the form", () => {
    let overmind: OvermindMock<Config>;

    beforeEach(() => {
      overmind = createTestOvermind(
        {},
        {
          login: {
            mode: {
              current: "registering"
            },
            username: "testuser"
          }
        }
      );
    });
    test("Submit without entering anything", async () => {
      const { container, getByText } = render(
        <Provider value={overmind}>
          <RegisterForm />
        </Provider>
      );
      await wait(() =>
        fireEvent.click(getByText("Register", { selector: "button" }))
      );
      expect(container).toMatchSnapshot();
    });
    test("Submit with whitespace for everything", async () => {
      const { container, getByText, getByLabelText } = render(
        <Provider value={overmind}>
          <RegisterForm />
        </Provider>
      );
      await wait(() => {
        fireEvent.change(getByLabelText("Email Address"), {
          target: { value: "   " }
        });
        fireEvent.change(getByLabelText("Display Name"), {
          target: { value: "   " }
        });
        fireEvent.change(getByLabelText("Password"), {
          target: { value: "   " }
        });
        fireEvent.change(getByLabelText("Repeat Password"), {
          target: { value: "   " }
        });
        fireEvent.click(getByText("Register", { selector: "button" }));
      });
      expect(container).toMatchSnapshot();
    });
    test("Submit with invalid email address", async () => {
      const { container, getByText, getByLabelText } = render(
        <Provider value={overmind}>
          <RegisterForm />
        </Provider>
      );
      await wait(() => {
        fireEvent.change(getByLabelText("Email Address"), {
          target: { value: "testuser" }
        });
        fireEvent.change(getByLabelText("Display Name"), {
          target: { value: "Test User" }
        });
        fireEvent.change(getByLabelText("Password"), {
          target: { value: "Pa55word" }
        });
        fireEvent.change(getByLabelText("Repeat Password"), {
          target: { value: "Pa55word" }
        });
        fireEvent.click(getByText("Register", { selector: "button" }));
      });
      expect(container).toMatchSnapshot();
    });
    test("Submit with mismatched passwords", async () => {
      const { container, getByText, getByLabelText } = render(
        <Provider value={overmind}>
          <RegisterForm />
        </Provider>
      );
      await wait(() => {
        fireEvent.change(getByLabelText("Email Address"), {
          target: { value: "testuser@example.com" }
        });
        fireEvent.change(getByLabelText("Display Name"), {
          target: { value: "Test User" }
        });
        fireEvent.change(getByLabelText("Password"), {
          target: { value: "Pa55word" }
        });
        fireEvent.change(getByLabelText("Repeat Password"), {
          target: { value: "password" }
        });
        fireEvent.click(getByText("Register", { selector: "button" }));
      });
      expect(container).toMatchSnapshot();
    });
  });
});
