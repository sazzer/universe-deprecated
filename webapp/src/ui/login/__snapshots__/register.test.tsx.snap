// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm Initial render 1`] = `
<div>
  <h3>
    Register
  </h3>
  <form
    data-test="registerForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="email"
    >
      <label
        for="email"
      >
        Email Address
      </label>
      <input
        class="form-control"
        id="email"
        name="email"
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="displayName"
    >
      <label
        for="displayName"
      >
        Display Name
      </label>
      <input
        class="form-control"
        id="displayName"
        name="displayName"
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        Password
      </label>
      <input
        class="form-control"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <div
      class="form-group"
      data-test="password2"
    >
      <label
        for="password2"
      >
        Repeat Password
      </label>
      <input
        class="form-control"
        id="password2"
        name="password2"
        type="password"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Register
      </button>
      <button
        class="btn btn-link"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;

exports[`RegisterForm Submitting the form Submit with invalid email address 1`] = `
<div>
  <h3>
    Register
  </h3>
  <form
    data-test="registerForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="email"
    >
      <label
        for="email"
      >
        Email Address
      </label>
      <input
        class="form-control is-invalid"
        id="email"
        name="email"
        type="text"
      />
      <div
        class="invalid-feedback"
      >
        Please enter a valid email address
      </div>
    </div>
    <div
      class="form-group"
      data-test="displayName"
    >
      <label
        for="displayName"
      >
        Display Name
      </label>
      <input
        class="form-control"
        id="displayName"
        name="displayName"
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        Password
      </label>
      <input
        class="form-control"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <div
      class="form-group"
      data-test="password2"
    >
      <label
        for="password2"
      >
        Repeat Password
      </label>
      <input
        class="form-control"
        id="password2"
        name="password2"
        type="password"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Register
      </button>
      <button
        class="btn btn-link"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;

exports[`RegisterForm Submitting the form Submit with mismatched passwords 1`] = `
<div>
  <h3>
    Register
  </h3>
  <form
    data-test="registerForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="email"
    >
      <label
        for="email"
      >
        Email Address
      </label>
      <input
        class="form-control"
        id="email"
        name="email"
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="displayName"
    >
      <label
        for="displayName"
      >
        Display Name
      </label>
      <input
        class="form-control"
        id="displayName"
        name="displayName"
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        Password
      </label>
      <input
        class="form-control"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <div
      class="form-group"
      data-test="password2"
    >
      <label
        for="password2"
      >
        Repeat Password
      </label>
      <input
        class="form-control is-invalid"
        id="password2"
        name="password2"
        type="password"
      />
      <div
        class="invalid-feedback"
      >
        Passwords do not match
      </div>
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Register
      </button>
      <button
        class="btn btn-link"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;

exports[`RegisterForm Submitting the form Submit with whitespace for everything 1`] = `
<div>
  <h3>
    Register
  </h3>
  <form
    data-test="registerForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="email"
    >
      <label
        for="email"
      >
        Email Address
      </label>
      <input
        class="form-control is-invalid"
        id="email"
        name="email"
        type="text"
      />
      <div
        class="invalid-feedback"
      >
        Please enter an email address
      </div>
    </div>
    <div
      class="form-group"
      data-test="displayName"
    >
      <label
        for="displayName"
      >
        Display Name
      </label>
      <input
        class="form-control is-invalid"
        id="displayName"
        name="displayName"
        type="text"
      />
      <div
        class="invalid-feedback"
      >
        Please enter a display name
      </div>
    </div>
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        Password
      </label>
      <input
        class="form-control"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <div
      class="form-group"
      data-test="password2"
    >
      <label
        for="password2"
      >
        Repeat Password
      </label>
      <input
        class="form-control"
        id="password2"
        name="password2"
        type="password"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Register
      </button>
      <button
        class="btn btn-link"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;

exports[`RegisterForm Submitting the form Submit without entering anything 1`] = `
<div>
  <h3>
    Register
  </h3>
  <form
    data-test="registerForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="email"
    >
      <label
        for="email"
      >
        Email Address
      </label>
      <input
        class="form-control is-invalid"
        id="email"
        name="email"
        type="text"
      />
      <div
        class="invalid-feedback"
      >
        Please enter an email address
      </div>
    </div>
    <div
      class="form-group"
      data-test="displayName"
    >
      <label
        for="displayName"
      >
        Display Name
      </label>
      <input
        class="form-control is-invalid"
        id="displayName"
        name="displayName"
        type="text"
      />
      <div
        class="invalid-feedback"
      >
        Please enter a display name
      </div>
    </div>
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        Password
      </label>
      <input
        class="form-control is-invalid"
        id="password"
        name="password"
        type="password"
      />
      <div
        class="invalid-feedback"
      >
        Please enter a password
      </div>
    </div>
    <div
      class="form-group"
      data-test="password2"
    >
      <label
        for="password2"
      >
        Repeat Password
      </label>
      <input
        class="form-control is-invalid"
        id="password2"
        name="password2"
        type="password"
      />
      <div
        class="invalid-feedback"
      >
        Please re-enter the password
      </div>
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Register
      </button>
      <button
        class="btn btn-link"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;
